include/range/v3/detail/bind_back.hpp

```diff
@@ -0,0 +1,113 @@
+/// \file
+// Range v3 library
+//
+//  Copyright Andrey Diduh 2019
+//
+//  Use, modification and distribution is subject to the
+//  Boost Software License, Version 1.0. (See accompanying
+//  file LICENSE_1_0.txt or copy at
+//  http://www.boost.org/LICENSE_1_0.txt)
+//
+// Project home: https://github.com/ericniebler/range-v3
+//
+
+#ifndef RANGE_V3_DETAIL_BIND_BACK_HPP
+#define RANGE_V3_DETAIL_BIND_BACK_HPP
+
+#include <functional>
+#include <tuple>
+#include <type_traits>
+#include <utility>
+
+#include <range/v3/range_fwd.hpp>
+
+#include <range/v3/functional/invoke.hpp>
+
+namespace ranges
+{
+    // bind_back like std::bind_front has no special treatment for nested
+    // bind-expressions. So there is no need to wrap Callables with ranges::protect.
+    // Also, bind_back does not unwrap std::ref's
+
+    namespace detail
+    {
+        template<int args_count, typename Fn, typename... Args>
+        class bind_back_t
+        {
+            using Self = bind_back_t<args_count, Fn, Args...>;
+
+            // clang-format off
+            template<typename Self, std::size_t... I, typename... CallArgs>
+            constexpr auto static CPP_auto_fun(apply)
+                (Self & self, std::index_sequence<I...>, CallArgs &&... call_args)
+            (
+                return self.fn(std::forward<CallArgs>(call_args)...,
```

Comment:
I'm unconvinced that this shouldn't be `invoke` here. I think for parity with `std::bind_front` it must.

Reply:
With `invoke` it will follow CURRENT `std::bind_front` contract.
With direct call - p1651.